docs/performance/productizing.rst

```diff
@@ -0,0 +1,116 @@
+.. _topics-to-watch-out-when-productizing-cratedb:
+
+###############################################
+ Topics to watch out when productizing CrateDB
+###############################################
+
+.. _mindful-of-memory:
+
+*******************************
+ Be mindful of memory capacity
+*******************************
+
+In CrateDB, operations requiring a working set like groupings, aggregations, and
+sorting are performed fully in memory without spilling over to disk.
+
+Sometimes you may have a query that leads to a sub-optimal execution plan
+requiring lots of memory. If you are coming to CrateDB from other database
+systems, your experience may be that these queries will proceed to run taking
+longer than required and impacting other workloads in the meanwhile. Sometimes
+this effect may be obvious if a query takes a lot of resources and runs for a
+long time, other times it may go unnoticed if a query that could complete in say
+100 milliseconds takes one hundred times longer, 10 seconds, but the users put
+up with it without reporting to you.
+
+In CrateDB if a query would require more heap memory than the interested nodes
+have available the query will fail with a particular type of error message that
+we call a `CircuitBreakerException`, this is a fail-fast approach, we can see
```

Comment:
Why is this a fail-fast approach? It fails once the memory is occupied at runtime, could happen very late while processing queries. The main point here is: preventing the node from dying with an OOM.

Reply:
> It fails once the memory is occupied at runtime

In some cases yes, but in other cases I believe it does an estimate and it returns with the `CircuitBreakerException` right away based on the estimate, maybe you could check this?

Reply:
Estimates can only be done once the data is processed. When using e.g. literals, the value is available already and the estimate can be done immediately.

Reply:
In what cases did you experience "immediate" CBE's?